Add native logger support to RocksJava

neilramaswamy opened the pull request:
## Overview

In this PR, we introduce support for setting the RocksDB native logger through Java. As mentioned in the discussion on the [Google Group discussion](https://groups.google.com/g/rocksdb/c/xYmbEs4sqRM/m/e73E4whJAQAJ), this work is primarily motivated by the  JDK 17 [performance regression in JNI thread attach/detach calls](https://bugs.openjdk.org/browse/JDK-8314859): the only existing RocksJava logging configuration call, `setLogger`, invokes the provided logger over the JNI.

## Changes

Specifically, these changes add support for the `devnull` and `stderr` native loggers. For the `stderr` logger, we add the ability to prefix every log with a `logPrefix`, so that it becomes possible know which database a particular log is coming from (if multiple databases are in use). The  API looks like the following:

```java
Options opts = new Options();

NativeLogger stderrNativeLogger = NativeLogger.newStderrLogger(
  InfoLogLevel.DEBUG_LEVEL, "[my prefix here]");
options.setLogger(stderrNativeLogger);

try (final RocksDB db = RocksDB.open(options, ...))  {...}

// Cleanup
stderrNativeLogger.close()
opts.close();
```

Note that the API to set the logger is the same, via `Options::setLogger` (or `DBOptions::setLogger`). However, it will set the RocksDB logger to be native when  the provided logger is an instance of `NativeLogger`.

## Testing

Two tests have been added in `NativeLoggerTest.java`. The first test creates both the `devnull` and `stderr` loggers, and sets them on the associated `Options`. However, to avoid polluting the testing output with logs from `stderr`, only the `devnull` logger is actually used in the test. The second test does the same logic, but for `DBOptions`.

It is possible to manually verify the `stderr` logger by modifying the tests slightly, and observing that the console indeed gets cluttered with logs from `stderr`.

neilramaswamy commented:
Hi @adamretter, I'd love if you could take a look at these updates to RocksJava. Happy to make any changes as needed.

adamretter commented:
@neilramaswamy I think we should stick with just `setLogger(...)` in the DBOptions (and friends) and not add a `setNativeLogger(...)`. We have two options I think:

1. Rename `Logger` to `AbstractLogger`, and add a `Logger` Interface (which `AbstractLogger` should implement), leave `setLogger(Logger)`.
2. Add an interface called `LoggerInterface` which `Logger` implements, and also modify `setLogger(Logger)` to `setLogger(LoggerInterface)`.

Option 1 may break some existing applications but gives a cleaner API, whilst Option 2 is less disruptive but leaves some cruft in the existing API design. I would be happy with either approach.

With either approach, the Native Loggers you are interested in can each be represented by a Java class which implements Logger (or an AbstractClass implementation thereof), and `setLogger()` can handle both native and non-native loggers by checking for their sub-class.

neilramaswamy commented:
@adamretter, thank you for your quick comments. I've implemented the second option, since I didn't want to break any existing code.

But yes, there certainly some cruft: the `LoggerInterface` is very thin, and it's less a logger _interface_ than a hack to let us pass either a `Logger` or `NativeLogger` to `setLogger`. But if you think it's cleaner than having `setNativeLogger`, I'm happy with this.

To make sure the cruft doesn't get in the way of usability, I can add documentation to the RocksJava Wiki page on how to create a JVM logger and set it, as well as configuring a native logger and setting that.

siying commented on util/stderr_logger.cc:
```diff
@@ -19,9 +19,10 @@ void StderrLogger::Logv(const char* format, va_list ap) {
+
+  fprintf(stderr, "%s%04d/%02d/%02d-%02d:%02d:%02d.%06d %llx ",
+          log_prefix.c_str(), t.tm_year + 1900, t.tm_mon + 1, t.tm_mday,
+          t.tm_hour, t.tm_min, t.tm_sec, static_cast<int>(now_tv.tv_usec),
```
I don't feel strongly about it, but I prefer timestamp is the first thing of the line, so that they are aligned when shown on screen. log_prefix can be after timestamp.

neilramaswamy replied:
This is reasonable. If you have multiple DBs (which the prefix assumes, anyway), a common prefix is certainly nicer.

adamretter commented:
@neilramaswamy I have pushed a couple of commits to refactor this more towards the API that I had in mind. I think perhaps that my previous explanation wasn't very detailed. Let me know if you are happy with it?

neilramaswamy commented on util/stderr_logger.h:
```diff
@@ -18,6 +18,8 @@ class StderrLogger : public Logger {
  public:
   explicit StderrLogger(const InfoLogLevel log_level = InfoLogLevel::INFO_LEVEL)
       : Logger(log_level) {}
+  explicit StderrLogger(const InfoLogLevel log_level, const std::string& prefix)
```
Yeah, that will be fine.

neilramaswamy commented on util/stderr_logger.h:
```diff
@@ -26,6 +28,10 @@ class StderrLogger : public Logger {
+
+ private:
+  // This prefix will be appended to the beginning of every log
+  std::string log_prefix;
```
Yeah, we can just store the c-str.

neilramaswamy commented on util/stderr_logger.cc:
```diff
@@ -19,10 +19,11 @@ void StderrLogger::Logv(const char* format, va_list ap) {
-          static_cast<int>(now_tv.tv_usec),
-          static_cast<long long unsigned int>(thread_id));
+
+  fprintf(stderr, "%04d/%02d/%02d-%02d:%02d:%02d.%06d %llx %s",
```
That's definitely a race, so I've updated this code to assemble the string in memory first. It's not the most optimized, but it is similar to what the JNI callback and Win loggers do today.

neilramaswamy commented:
This API works with me, but there will then end up being some redundancy in the `LoggerInterface` methods among the native loggers. Since there are very few more native loggers, this is probably fine.

adamretter reviewed (approved):
LGTM. Thanks :-)

adamretter commented:
> but there will then end up being some redundancy in the LoggerInterface methods among the native loggers

We can add an abstract sub-class in-future if needed I think @neilramaswamy

adamretter commented:
@siying @ajkr If you are happy with this, then I am happy with this now.

ajkr commented on util/stderr_logger.cc:
```diff
@@ -19,12 +19,36 @@ void StderrLogger::Logv(const char* format, va_list ap) {
+  // 36 bytes, but we allocate 40 to be safe.
+  size_t context_len = 40;
+
+  va_list ap_copy;
+  va_copy(ap_copy, ap);
+  const size_t log_suffix_len = vsnprintf(nullptr, 0, format, ap_copy);
+  va_end(ap_copy);
+
+  // Allocate space for the context, log_prefix, and log itself
+  // Extra byte for null termination
+  size_t buf_len = context_len + log_prefix_len + log_suffix_len + 1;
+  std::unique_ptr<char[]> buf(new char[buf_len]);
```
Neat

ajkr reviewed (changes requested):
The test failures here look relevant: https://github.com/facebook/rocksdb/actions/runs/7523821680/job/20477739549, https://github.com/facebook/rocksdb/actions/runs/7523821680/job/20477735891. Besides that and the inline comment, LGTM

neilramaswamy commented:
Thanks for the review @ajkr! I addressed the first issue, and I _think_ I fixed the test failure (they were only Windows build errors; I unfortunately do not have a Windows machine, so I can't test locally). I will wait for CI and take it from there.

ajkr reviewed (approved):
LGTM, thanks!

neilramaswamy commented:
Hi @ajkr, I believe you need to approve the workflow for all the checks to run (that's what GitHub tells me). I can take a look at the Windows build errors after that.

Also, I can't see the "Facebook Internal - Linter" check, which is failing. If you are able to share the warnings with me, I can fix them.

ajkr commented:
> Also, I can't see the "Facebook Internal - Linter" check, which is failing. If you are able to share the warnings with me, I can fix them.

The linter wants `StderrLogger`'s constructor to take the `prefix` parameter by const-reference (I know this was discussed already and don't really care either way). It also asks for the semicolon after that same constructor's (empty) body to be dropped

neilramaswamy commented:
Alas, all tests must run again, but I removed the semi. I think passing the string directly is more clear, so I left it. The Windows errors were fixed, so this should be the last time you have to approve the workflow :)

Thanks for the engagement on this. I really appreciate it.

cbi42 commented on util/stderr_logger.cc:
```diff
@@ -19,12 +19,36 @@ void StderrLogger::Logv(const char* format, va_list ap) {
+  int written =
+      snprintf(buf.get(), context_len + log_prefix_len,
+               // Keep this string in sync with context_len
+               "%04d/%02d/%02d-%02d:%02d:%02d.%06d %llx %s", t.tm_year + 1900,
```
Why is context_len 40, does thread_id not take 16 chars?

neilramaswamy replied:
The `thread_id` gets cast into a `long long unsigned int`. To be explicit, I got 40 because:

`int/int/int-int:int:int.int longlong` is 4 + 4 + 4 + 4 + 4 + 4 + 4 + 8 = 36.

Am I missing something?

cbi42 replied:
I got:
```
%04d/%02d/%02d-%02d:%02d:%02d.%06d %llx %s
4 + 1+2+ 1+2 +1+2+ 1+2+ 1+2+ 1+ 6 + 16+1  = 43
```

neilramaswamy replied:
Yeah, this is closer—I forgot to count lengths. But actually, I think the 43 is one too little, I think you forgot the space before the `%llx`. So I think it is actually 44. Since this seems to be very error prone, I have included a comment with counting in the source.
